File: Sources/OpenPocketViewCore/FeedPresentPolicy.swift
import Foundation

/// Present-path hygiene for live and playback.
///
/// Freeze is a flag, not a flush. Latest frame wins. A hidden Metal / GLES
/// drawable is a black well — unhide replace-grade *before* `nextDrawable`.
/// Overlay chrome may stay hidden until the bake lands so an opaque empty
/// layer cannot cover identity.
///
/// Shells own the GPU. This type is the shared predicate.
public enum FeedPresentPolicy: Sendable {
    /// No new present for this long is a freeze. Matches ``FeedWatchdog/stallThreshold``.
    /// UDP still alive + frozen canvas is a present hitch, not a recover enable.
    public static let freezeThreshold: TimeInterval = FeedWatchdog.stallThreshold

    /// Runtime LUT / peaking / zebra working width. Cubes are resolution-independent;
    /// convolution graphs must not run at a 4K original when the monitor is 720p.
    public static let maxWorkingWidth: Int = 1440

    /// Keep the current-orientation picture this many presents before extra-mirror
    /// X-flips. Instant scaleX on the last frame is the jarring flip-flip.
    public static let extraMirrorHoldFrames: Int = 3
    /// Cap so a stalled decoder cannot freeze extra-mirror. 3 frames at 25 fps is 120 ms.
    public static let extraMirrorHoldSeconds: TimeInterval = 0.2

    /// `true` while the last picture should stay on screen (old transform).
    public static func shouldHoldPictureAcrossMirror(
        framesHeld: Int, secondsHeld: TimeInterval
    ) -> Bool {
        framesHeld <= extraMirrorHoldFrames && secondsHeld < extraMirrorHoldSeconds
    }

    /// Visible, enabled, attached view with a real drawable. Overlay may be hidden
    /// until the first bake — use ``shouldScheduleBake(enabled:hasDrawable:)`` there.
    public static func shouldRender(
        attached: Bool,
        enabled: Bool,
        hidden: Bool,
        hasDrawable: Bool
    ) -> Bool {
        attached && enabled && !hidden && hasDrawable
    }

    /// Schedule a bake. Hidden overlay is OK; replace unhides before the drawable.
    public static func shouldScheduleBake(enabled: Bool, hasDrawable: Bool) -> Bool {
        enabled && hasDrawable
    }

    /// MTKView / display-link skip: same timestamp as the last successful present.
    /// `timeNs == 0` means unknown — never skip.
    public static func isDuplicateFrameTime(_ timeNs: Int64, lastPresentedNs: Int64) -> Bool {
        timeNs != 0 && lastPresentedNs != 0 && timeNs == lastPresentedNs
    }

    public static func isFrozen(secondsSinceLastPresent: TimeInterval?) -> Bool {
        guard let age = secondsSinceLastPresent else { return false }
        return age >= freezeThreshold
    }

    /// Overlay bakes set `hasPresentedFrame`, but LUT must not treat that as
    /// replace-ownership (opaque Metal over identity).
    public static func replaceOwnsPicture(
        hasPresentedFrame: Bool,
        lastPresentWasOverlay: Bool
    ) -> Bool {
        hasPresentedFrame && !lastPresentWasOverlay
    }

    /// Unhide the processed feed before `nextDrawable` on replace-grade.
    /// Overlay stays hidden until the transparent bake lands.
    public static func unhideMetalBeforeBake(overlay: Bool) -> Bool {
        !overlay
    }

    /// Prefer the 720p LRF/XRF proxy for monitor grade. 4K original is export-only.
    public static func preferProxyForMonitorGrade(hasProxy: Bool) -> Bool {
        hasProxy
    }

    /// Keep the last sample through recover. Flush is disconnect or a failed layer
    /// that already has a replacement picture, never a stall.
    public static func shouldFlushDisplayedImage(
        disconnecting: Bool,
        layerFailed: Bool,
        nextFrameReady: Bool
    ) -> Bool {
        if disconnecting { return true }
        if layerFailed { return nextFrameReady }
        return false
    }
}

/// One live-enable / recover write in flight. Overlapping `0x09/0xa8` is a black well.
public struct SerialSessionGate: Equatable, Sendable {
    public private(set) var inFlight = false

    public init() {}

    /// `true` if this caller owns the slot. Pair with ``end()``.
    public mutating func begin() -> Bool {
        guard !inFlight else { return false }
        inFlight = true
        return true
    }

    public mutating func end() {
        inFlight = false
    }
}

/// Delays extra-mirror so the on-screen orientation is not X-flipped in place.
public struct ExtraMirrorHold: Equatable, Sendable {
    public enum Step: Equatable, Sendable {
        case unchanged
        case hold
        case commit(Bool)
    }

    public private(set) var displayed: Bool?
    public private(set) var pending: Bool?
    public private(set) var framesHeld = 0
    public private(set) var startedAt: TimeInterval?

    public init() {}

    public mutating func reset() {
        displayed = nil
        pending = nil
        framesHeld = 0
        startedAt = nil
    }

    /// First present commits immediately. A later edge holds
    /// ``FeedPresentPolicy/extraMirrorHoldFrames`` then commits.
    public mutating func step(want: Bool, now: TimeInterval) -> Step {
        if displayed == nil {
            displayed = want
            pending = nil
            framesHeld = 0
            startedAt = nil
            return .commit(want)
        }
        let shown = displayed!
        if pending == nil {
            guard want != shown else { return .unchanged }
            pending = want
            framesHeld = 0
            startedAt = now
        } else if pending != want {
            if want == shown {
                pending = nil
                framesHeld = 0
                startedAt = nil
                return .unchanged
            }
            pending = want
            framesHeld = 0
            startedAt = now
        }
        framesHeld += 1
        let age = now - (startedAt ?? now)
        if FeedPresentPolicy.shouldHoldPictureAcrossMirror(
            framesHeld: framesHeld, secondsHeld: age)
        {
            return .hold
        }
        displayed = want
        pending = nil
        framesHeld = 0
        startedAt = nil
        return .commit(want)
    }
}

File: Apps/Android/app/src/main/kotlin/com/opencapture/openpocketcine/feed/FeedPresentPolicy.kt
package com.opencapture.openpocketcine.feed

/**
 * OpenPocketViewCore `FeedPresentPolicy` — present-path hygiene.
 *
 * Freeze is a flag, not a flush. Latest frame wins. A hidden GLES window is a
 * black well. Keep in lockstep with the Swift tests.
 */
object FeedPresentPolicy {
    const val FREEZE_THRESHOLD_SECONDS = 2.0
    const val MAX_WORKING_WIDTH = 1440
    const val EXTRA_MIRROR_HOLD_FRAMES = 3
    const val EXTRA_MIRROR_HOLD_SECONDS = 0.2
    /** 3 frames at 25 fps. Compose delay while the last orientation stays on screen. */
    const val EXTRA_MIRROR_HOLD_MS = 120L

    /**
     * iOS `FeedWorkingRaster.targetSize`: aspect-preserving cap on the graded raster.
     * 720p proxies pass through; a 4K original grades at 1440 wide.
     */
    fun workingSize(width: Int, height: Int, maxWidth: Int = MAX_WORKING_WIDTH): Pair<Int, Int> {
        if (width <= maxWidth || width <= 1 || height <= 1) return width to height
        return maxWidth to maxOf(1, Math.round(height * maxWidth.toDouble() / width).toInt())
    }

    /** Copy-pass tap offset, in destination pixels, for a downsample of 1.5x or more. */
    fun downsampleSpread(sourceWidth: Int, targetWidth: Int): Float =
        if (targetWidth > 0 && sourceWidth * 2 >= targetWidth * 3) 0.25f else 0f

    fun shouldHoldPictureAcrossMirror(framesHeld: Int, secondsHeld: Double): Boolean =
        framesHeld <= EXTRA_MIRROR_HOLD_FRAMES && secondsHeld < EXTRA_MIRROR_HOLD_SECONDS

    fun shouldRender(
        attached: Boolean,
        enabled: Boolean,
        hidden: Boolean,
        hasDrawable: Boolean,
    ): Boolean = attached && enabled && !hidden && hasDrawable

    fun shouldScheduleBake(enabled: Boolean, hasDrawable: Boolean): Boolean = enabled && hasDrawable

    fun isDuplicateFrameTime(timeNs: Long, lastPresentedNs: Long): Boolean =
        timeNs != 0L && lastPresentedNs != 0L && timeNs == lastPresentedNs

    fun isFrozen(secondsSinceLastPresent: Double?): Boolean {
        val age = secondsSinceLastPresent ?: return false
        return age >= FREEZE_THRESHOLD_SECONDS
    }

    fun replaceOwnsPicture(hasPresentedFrame: Boolean, lastPresentWasOverlay: Boolean): Boolean =
        hasPresentedFrame && !lastPresentWasOverlay

    fun unhideMetalBeforeBake(overlay: Boolean): Boolean = !overlay

    fun preferProxyForMonitorGrade(hasProxy: Boolean): Boolean = hasProxy

    fun shouldFlushDisplayedImage(
        disconnecting: Boolean,
        layerFailed: Boolean,
        nextFrameReady: Boolean,
    ): Boolean {
        if (disconnecting) return true
        if (layerFailed) return nextFrameReady
        return false
    }
}

/** One live-enable write in flight. Overlapping `0x09/0xa8` is a black well. */
class SerialSessionGate {
    var inFlight: Boolean = false
        private set

    fun begin(): Boolean {
        if (inFlight) return false
        inFlight = true
        return true
    }

    fun end() {
        inFlight = false
    }
}

/** Delays extra-mirror so the on-screen orientation is not X-flipped in place. */
class ExtraMirrorHold {
    enum class Step {
        UNCHANGED,
        HOLD,
        COMMIT,
    }

    data class Result(val step: Step, val mirrored: Boolean)

    var displayed: Boolean? = null
        private set
    private var pending: Boolean? = null
    private var framesHeld = 0
    private var startedAt: Double? = null

    fun reset() {
        displayed = null
        pending = null
        framesHeld = 0
        startedAt = null
    }

    fun step(want: Boolean, now: Double): Result {
        val shown = displayed
        if (shown == null) {
            displayed = want
            pending = null
            framesHeld = 0
            startedAt = null
            return Result(Step.COMMIT, want)
        }
        if (pending == null) {
            if (want == shown) return Result(Step.UNCHANGED, shown)
            pending = want
            framesHeld = 0
            startedAt = now
        } else if (pending != want) {
            if (want == shown) {
                pending = null
                framesHeld = 0
                startedAt = null
                return Result(Step.UNCHANGED, shown)
            }
            pending = want
            framesHeld = 0
            startedAt = now
        }
        framesHeld += 1
        val age = now - (startedAt ?: now)
        if (FeedPresentPolicy.shouldHoldPictureAcrossMirror(framesHeld, age)) {
            return Result(Step.HOLD, shown)
        }
        displayed = want
        pending = null
        framesHeld = 0
        startedAt = null
        return Result(Step.COMMIT, want)
    }
}
